frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/traverse/BaseTraverseAction.java

```diff
@@ -0,0 +1,129 @@
+    val gitRepository = getSelectedGitRepository(anActionEvent);
+    val graphTable = getGraphTable(anActionEvent);
+    val repositorySnapshot = graphTable.getGitMacheteRepositorySnapshot();
+    val branchLayout = repositorySnapshot != null ? repositorySnapshot.getBranchLayout() : null;
+    val project = getProject(anActionEvent);
+
+    boolean yesNoResult = true;
+    if (branchLayout != null && branchLayout.getRootEntries().nonEmpty() && gitRepository != null) {
+      if (PropertiesComponent.getInstance(project).getBoolean(SHOW_TRAVERSE_INFO, /* defaultValue */ true)) {
+
+        val traverseInfoDialog = new InfoDialog(
+            project,
+            getString("action.GitMachete.BaseTraverseAction.dialog.traverse-approval.title"),
+            getString("action.GitMachete.BaseTraverseAction.dialog.traverse-approval.text.HTML"),
+            SHOW_TRAVERSE_INFO,
+            /* myHeight */ 100);
+
+        yesNoResult = traverseInfoDialog.showAndGet();
+      }
+
+      if (yesNoResult) {
+        val entryName = getNameOfBranchUnderAction(anActionEvent);
+        if (entryName != null) {
+          checkoutAndTraverse(gitRepository, graphTable, entryName);
+        } else {
+          LOG.warn("Skipping traverse action because entryName is undefined");
+        }
+      }
+    }
+  }
+
+  private void checkoutAndTraverse(GitRepository gitRepository, BaseEnhancedGraphTable graphTable, String branchName) {
+    log().debug(() -> "Queuing '${branchName}' branch checkout background task");
+
+    @UI Runnable traverseRunnable = () -> traverse(gitRepository, graphTable, branchName);
+    Runnable repositoryRefreshRunnable = () -> graphTable.queueRepositoryUpdateAndModelRefresh(traverseRunnable);
+    Runnable callInAwtLater = () -> ModalityUiUtil.invokeLaterIfNeeded(NON_MODAL, repositoryRefreshRunnable);
+    val gitBrancher = GitBrancher.getInstance(gitRepository.getProject());
+    val repositories = Collections.singletonList(gitRepository);
+    gitBrancher.checkout(/* reference */ branchName, /* detach */ false, repositories, callInAwtLater);
+  }
+
+  @UIEffect
+  private void traverse(GitRepository gitRepository, BaseEnhancedGraphTable graphTable, String branchName) {
+    val repositorySnapshot = graphTable.getGitMacheteRepositorySnapshot();
+    val branchLayout = repositorySnapshot != null ? repositorySnapshot.getBranchLayout() : null;
+    val gitMacheteBranch = repositorySnapshot != null ? repositorySnapshot.getManagedBranchByName(branchName) : null;
+
+    if (gitMacheteBranch != null) {
+      @UI Runnable traverseNextEntry = () -> {
+        var nextBranch = branchLayout != null ? branchLayout.findNextEntry(branchName) : null;
+        if (nextBranch != null) {
+          checkoutAndTraverse(gitRepository, graphTable, nextBranch.getName());
+        }
+      };
+
+      if (gitMacheteBranch.isNonRoot()) {
+        assert repositorySnapshot != null : "repositorySnapshot is null";
```

Comment:
Huh Checker should get that `repositorySnapshot` can't be null here... anyway, rather than putting a assertion, do sth like that at the  beginning of the method:

```java
val repositorySnapshot = graphTable.getGitMacheteRepositorySnapshot();
if (repositorySnapshot == null) return;
```

Reply:
for me:
`assert smth != null` indicates that we can be sure that the value is not null 
`if (smth == null) return` indicates that smth can be null and we simply end at this point in the method

I would rather leave the assert here - from the code reader POV assert makes me thinking "why" and then I go to the point and see why. on the other hand seeing `if smth  == null return` I do not think much - assuming thats possible and expected in some cases

Reply:
Okay... but at the **beginning** of the method, it's perfectly expected that `repositorySnapshot` might be null, right? 🤔 if we return in such case, then all of:

1. `val branchLayout = repositorySnapshot != null ? repositorySnapshot.getBranchLayout() : null;`
2. `val gitMacheteBranch = repositorySnapshot != null ? repositorySnapshot.getManagedBranchByName(branchName) : null;`
3. `assert repositorySnapshot != null : "repositorySnapshot is null";`

can be simplified (1, 2) or straight removed (3)

Reply:
oh, ok you are right!